storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp

```diff
@@ -0,0 +1,405 @@
+        // If previous reconnection attempts have failed then
+        // restart the reconnection process
+        LOG_DEBUG("GetNdbObject triggered reconnection");
+        Reconnect();
+      }
+
+      LOG_ERROR(ERROR_033 + std::string(" Connection State: ") + std::to_string(connection_state) +
+                std::string(" Reconnection State: ") + std::to_string(reconnection_in_progress));
+      return RS_SERVER_ERROR(ERROR_033);
+    }
+  }
+
+  {
+    std::lock_guard<std::mutex> guardInfo(connectionInfoMutex);
+    std::lock_guard<std::mutex> guard(connectionMutex);
+    RS_Status ret_status = RS_OK;
+    if (availableNdbObjects.empty()) {
+      *ndb_object = new Ndb(ndbConnection);
+      int retCode = (*ndb_object)->init();
+      if (retCode != 0) {
+        delete ndb_object;
+        ret_status =
+            RS_SERVER_ERROR(ERROR_004 + std::string(" RetCode: ") + std::to_string(retCode));
+      }
+      __atomic_fetch_add(&stats.ndb_objects_created, 1, __ATOMIC_SEQ_CST);
+      __atomic_fetch_add(&stats.ndb_objects_count, 1, __ATOMIC_SEQ_CST);
+      allAvailableNdbObjects.push_back(*ndb_object);
+    } else {
+      *ndb_object = availableNdbObjects.front();
+      availableNdbObjects.pop_front();
+    }
+    return ret_status;
+  }
+}
+
+//--------------------------------------------------------------------------------------------------
+
+void RDRSRonDBConnection::ReturnNDBObjectToPool(Ndb *ndb_object, RS_Status *status) {
+  {
+    std::lock_guard<std::mutex> guard(connectionMutex);
+    availableNdbObjects.push_back(ndb_object);
+  }
+
+  // Note there are not unit test for this
+  // Inorder to test this run the  TestReconnection1 for longer duration
+  // and then drop the ndbconnection  using iptables or by disconnection the network
+  if (status != nullptr && status->http_code != SUCCESS) {  // check for errors
+    // Classification.UnknownResultError is the classification
+    // for loss of connectivity to the cluster
+    if (status->classification == NdbError::UnknownResultError) {
+      LOG_ERROR("Detected connection loss. Triggering reconnection.");
+      Reconnect();
+    }
+  }
+}
+
+//--------------------------------------------------------------------------------------------------
+
+RonDB_Stats RDRSRonDBConnection::GetStats() {
+  std::lock_guard<std::mutex> guardInfo(connectionInfoMutex);
+
+  stats.ndb_objects_available = availableNdbObjects.size();
+  return stats;
+}
+
+//--------------------------------------------------------------------------------------------------
+RS_Status RDRSRonDBConnection::Shutdown() {
+  return Shutdown(false);
+}
+
+//--------------------------------------------------------------------------------------------------
+
+RS_Status RDRSRonDBConnection::Shutdown(bool end) {
+
+  LOG_DEBUG("Shutting down RonDB connection and NDB object pool");
+
+  {
+    std::lock_guard<std::mutex> guardInfo(connectionInfoMutex);
+    stats.connection_state = DISCONNECTED;
+  }
+
```

Comment:
Again release of mutex immediately followed by new seize of mutex.

Reply:
Or actually not

Reply:
refactored